src/gpu/kmeans/kmeans_impl.h

```diff
@@ -212,34 +217,21 @@ int kmeans(
 
     // whether to perform per iteration check
     if (do_per_iter_check) {
-      double distance_sum = 0.0;
-      int moved_points = 0.0;
-      for (int q = 0; q < n_gpu; q++) {
-        safe_cuda(cudaSetDevice(dList[q])); //  unnecessary
-        safe_cuda(cudaMemcpyAsync(h_changes + q, d_changes[q], sizeof(int), cudaMemcpyDeviceToHost, cuda_stream[q]));
-        safe_cuda(cudaMemcpyAsync(h_distance_sum + q,
-                                  d_distance_sum[q],
-                                  sizeof(T),
-                                  cudaMemcpyDeviceToHost,
-                                  cuda_stream[q]));
-        detail::streamsync(dList[q]);
-        if (verbose >= 2) {
-          std::cout << "Device " << dList[q] << ":  Iteration " << i << " produced " << h_changes[q]
-                    << " changes and the total_distance is " << h_distance_sum[q] << std::endl;
-        }
-        distance_sum += h_distance_sum[q];
-        moved_points += h_changes[q];
-      }
-      if (i > 0) {
-        double fraction = (double) moved_points / n;
-#define NUMSTEP 10
-        if (verbose > 1 && (i <= 1 || i % NUMSTEP == 0)) {
-          std::cout << "Iteration: " << i << ", moved points: " << moved_points << std::endl;
-        }
-        if (fraction < threshold) {
-          if (verbose) { std::cout << "Threshold triggered. Terminating early." << std::endl; }
-          done = true;
-        }
+      safe_cuda(cudaSetDevice(dList[0]));
+      thrust::device_vector<T> diff(d_old_centroids.size());
+
+      thrust::transform(
+          d_old_centroids.begin(), d_old_centroids.end(),
+          (*centroids[0]).begin(),
+          diff.begin(),
+          thrust::minus<T>()
+      );
+      thrust::device_vector<T> squared_norm(1);
+      kmeans::detail::make_self_dots(1, k * d, diff, squared_norm);
```

Comment:
Consider using thrust inner product instead of make self dots: https://thrust.github.io/doc/group__transformed__reductions.html

This can actually be fused with the previous transform operation by defining a binary operator as (a-b)^2. Then you don't need the diff array. 

Remind me to also show you how to define a custom allocator for thrust at some point. This is important for performance reasons when calling certain thrust functions in a loop e.g. (reduction, inner product).

Reply:
@RAMitchell sweet, thanks!

Are you coming to MV for H2O World? Could start making a list of tips-and-tricks you could show us live then :-)